esp-hal/src/rmt.rs

```diff
@@ -2418,3 +2418,128 @@ mod chip_specific {
     pub(crate) use impl_rx_channel;
     pub(crate) use impl_tx_channel;
 }
+
+#[cfg(not(any(esp32, esp32s2)))]
+mod clock_divider_solver {
+    // Calculate the greatest common divisor of a and b
+    const fn gcd(mut a: u32, mut b: u32) -> u32 {
+        while b != 0 {
+            (a, b) = (b, (a % b))
+        }
+        a
+    }
+
+    pub struct ClockDivider {
+        // Integral clock divider value.
+        pub div_num: u32,
+
+        // Fractional clock divider numerator value.
+        pub div_b: u32,
+
+        // Fractional clock divider denominator value.
+        pub div_a: u32,
+    }
+
+    impl ClockDivider {
+        pub const fn from_frequence(
+            source: u32,
+            target: u32,
+            integral_max: u32,
+            fractional_max: u32,
+        ) -> Option<Self> {
+            if target > source {
+                return None;
+            }
+
+            let quotient = source / target;
+            if quotient > integral_max {
+                return None;
+            }
+
+            let remainder = source % target;
+            if remainder == 0 {
+                return Some(Self {
+                    div_num: quotient,
+                    div_b: 0,
+                    div_a: 0,
+                });
+            }
+
+            let gcd = gcd(target, remainder);
+            let numerator = remainder / gcd;
+            let denominator = target / gcd;
+            if numerator <= fractional_max && denominator <= fractional_max {
+                return Some(Self {
+                    div_num: quotient,
+                    div_b: numerator,
+                    div_a: denominator,
+                });
+            }
+
+            // Search Stern-Brocot Tree
+            let target_frac = Fraction {
+                numerator,
+                denominator,
+            };
+            let mut l = Fraction {
+                numerator: 0,
+                denominator: 1,
+            };
+            let mut h = Fraction {
+                numerator: 1,
+                denominator: 1, // We only search the left harf part
+            };
+            loop {
+                let m = Fraction {
+                    numerator: l.numerator + h.numerator,
+                    denominator: l.denominator + h.denominator,
+                };
+                if m.numerator > fractional_max || m.denominator > fractional_max {
+                    // L and H, which is closer?
+
+                    // M - L
+                    let a = Fraction {
```

Comment:
Let's define these operations on `Fraction`, that way the code can explain what it does instead of you commenting `M - L` on top.

Reply:
Then we can't make the function const.

Reply:
Or we just define our `const fn sub()` and `const fn partial_cmp()`, not using the `Sub` and `PartialOrd` trait.

Reply:
I'm fine with custom const fns, this isn't a public type that has to be as generic as possible.